runtime/src/iree/base/queue_test.cc

```diff
@@ -0,0 +1,110 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/base/queue.h"
+
+#include "iree/testing/gtest.h"
+
+IREE_TYPED_QUEUE_WRAPPER(test_queue, int32_t, 4);
+
+class QueueTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    test_queue_initialize(&queue_, iree_allocator_system());
+  }
+
+  void TearDown() override { test_queue_deinitialize(&queue_); }
+
+  test_queue_t queue_;
+};
+
+TEST_F(QueueTest, initialize) {
+  EXPECT_EQ(queue_.element_count, 0);
+  EXPECT_EQ(queue_.capacity, 4);
+  EXPECT_EQ(queue_.element_size, sizeof(int));
+}
+
+TEST_F(QueueTest, push_back) {
+  int value = 42;
+  IREE_CHECK_OK(test_queue_push_back(&queue_, value));
+  EXPECT_EQ(queue_.element_count, 1);
+  EXPECT_EQ(test_queue_at(&queue_, 0), value);
+}
+
+TEST_F(QueueTest, push_back_and_expand) {
+  for (int i = 0; i < 5; ++i) {
+    IREE_CHECK_OK(test_queue_push_back(&queue_, i));
+  }
+  EXPECT_EQ(queue_.element_count, 5);
+  EXPECT_GT(queue_.capacity, 4);
+}
+
+TEST_F(QueueTest, pop_front) {
+  IREE_CHECK_OK(test_queue_push_back(&queue_, 12));
+  IREE_CHECK_OK(test_queue_push_back(&queue_, 15));
+  test_queue_pop_front(&queue_, 1);
+  EXPECT_EQ(queue_.element_count, 1);
+  EXPECT_EQ(test_queue_at(&queue_, 0), 15);
+}
+
+TEST_F(QueueTest, at) {
+  int value1 = 42;
+  int value2 = 84;
+  IREE_CHECK_OK(test_queue_push_back(&queue_, value1));
+  IREE_CHECK_OK(test_queue_push_back(&queue_, value2));
+  EXPECT_EQ(test_queue_at(&queue_, 0), value1);
+  EXPECT_EQ(test_queue_at(&queue_, 1), value2);
+}
+
+TEST_F(QueueTest, cycle_around_queue) {
+  for (int i = 0; i < 4; ++i) {
+    IREE_CHECK_OK(test_queue_push_back(&queue_, i));
+  }
+  for (int i = 0; i < 4; ++i) {
+    test_queue_pop_front(&queue_, 1);
+    IREE_CHECK_OK(test_queue_push_back(&queue_, i + 4));
+  }
+  for (int i = 0; i < 4; ++i) {
+    EXPECT_EQ(test_queue_at(&queue_, i), i + 4);
+  }
+  EXPECT_EQ(queue_.element_count, 4);
+}
+
+TEST_F(QueueTest, cycle_around_queue_twice) {
+  for (int i = 0; i < 4; ++i) {
+    IREE_CHECK_OK(test_queue_push_back(&queue_, i));
+  }
+  for (int i = 0; i < 8; ++i) {
+    test_queue_pop_front(&queue_, 1);
+    IREE_CHECK_OK(test_queue_push_back(&queue_, i + 4));
+  }
+  for (int i = 0; i < 4; ++i) {
+    EXPECT_EQ(test_queue_at(&queue_, i), i + 8);
+  }
+  EXPECT_EQ(queue_.element_count, 4);
+}
+
+TEST_F(QueueTest, allocate_twice) {
+  for (int i = 0; i < 8; ++i) {
+    IREE_CHECK_OK(test_queue_push_back(&queue_, i));
+  }
+  EXPECT_EQ(queue_.element_count, 8);
+  EXPECT_GT(queue_.capacity, 4);
+
+  for (int i = 8; i < 16; ++i) {
+    IREE_CHECK_OK(test_queue_push_back(&queue_, i));
```

Comment:
IREE_CHECK_OK is an abort (and we try never to use it) - for tests you want IREE_EXPECT_OK or IREE_ASSERT_OK (which are gtest EXPECT/ASSERT): 
https://google.github.io/googletest/reference/assertions.html

tldr: expect is nonfatal if you want more useful errors from the same test - like a print of both element_count and capacity being wrong below, while assert is fatal and will stop the test immediately - like if an allocation/initialization fails and every condition after is guaranteed to fail/crash.

for testing specific non-OK status codes we have a helper that uses the matcher style: IREE_EXPECT_STATUS_IS/IREE_ASSERT_STATUS_IS - those will print the full status message if it does not match